### DISCUSSION TITLE
Is it possible to support USGS 3DEP maps as underlay?

### DISCUSSION THREAD

**Discussion by endolith** [osmandapp/OsmAnd#17256] [Q&A] [answered]
These maps are much better at showing cliff edges for rock climbing:

![image](https://github.com/osmandapp/OsmAnd/assets/58611/245a0a80-c818-4d47-af9f-eb73d05520ff)

![image](https://github.com/osmandapp/OsmAnd/assets/58611/4bab4f40-7ea3-404f-8181-50a70b2bf4c1)

I'm using this URL to view them in iD editor for OpenStreetMap:

`https://elevation.nationalmap.gov/arcgis/services/3DEPElevation/ImageServer/WMSServer?FORMAT=image/png32&TRANSPARENT=TRUE&VERSION=1.3.0&SERVICE=WMS&REQUEST=GetMap&LAYERS=3DEPElevation:Hillshade%20Multidirectional&STYLES=&CRS={proj}&WIDTH={width}&HEIGHT={height}&BBOX={bbox}`

Is there a way to view these tiles in OsmAnd too?

ChatGPT said this URL format would work: `https://elevation.nationalmap.gov/arcgis/services/3DEPElevation/ImageServer/WMSServer?FORMAT=image/png32&TRANSPARENT=TRUE&VERSION=1.3.0&SERVICE=WMS&REQUEST=GetMap&LAYERS=3DEPElevation:Hillshade%20Multidirectional&STYLES=&CRS=EPSG:3857&WIDTH=256&HEIGHT=256&BBOX={bbox}`  But it seems to do nothing, the background is just grey.

It said I should check https://elevation.nationalmap.gov/arcgis/services/3DEPElevation/ImageServer/WMSServer?request=GetCapabilities&service=WMS to see if it supports EPSG:3857, and it seems to:

```xml
…
<Layer>
<Name>0</Name>
<Title>3DEPElevation</Title>
<Abstract>
The USGS 3D Elevation Program (3DEP) Bare Earth DEM Dynamic service is based on multi-resolution USGS DEM sources and provides dynamic functions for visualization. These functions include: Hillshade, Aspect Map, Hillshade Stretched, Multi-directional Hillshade, Slope Map, Elevation Tinted Hillshade, Contour. In addition the OGC Web Map Service (WMS) and Web Coverage Service (WCS) interfaces are enabled. Data available in this map service reflects all 3DEP DEM data published as of May 15, 2023.
</Abstract>
<CRS>CRS:84</CRS>
<CRS>EPSG:4326</CRS>
<CRS>EPSG:3857</CRS>
…
<EX_GeographicBoundingBox>
<westBoundLongitude>-179.999989</westBoundLongitude>
<eastBoundLongitude>179.999993</eastBoundLongitude>
<southBoundLatitude>-15.001663</southBoundLatitude>
<northBoundLatitude>84.001679</northBoundLatitude>
</EX_GeographicBoundingBox>
…
<BoundingBox CRS="EPSG:3857" minx="-20037507.067200" miny="-1689391.823360" maxx="20037507.532527" maxy="18809001.971900"/>
…
```

**Comment by vshcherb** [comment]
Only if you export them somehow and make XYZ tiles, I guess.

**Comment by endolith** [comment]
https://github.com/osmandapp/Osmand/issues/4201#issuecomment-1697962680 😫

**Comment by endolith** [comment]
With an enormous amount of help from ChatGPT, I got a server set up on AWS, based on https://github.com/mapbox/whoots-js but hardcoding jpeg instead of png, and proxying the image instead of redirecting.  Here's my fork:

https://github.com/endolith/whoots-js

and the changed files:

https://github.com/mapbox/whoots-js/compare/master...endolith:whoots-js:master

If anyone understand AWS billing and can explain to me how much it would cost to share it, I would consider sharing the server.  I think it's under the Free Tier for 12 months?

**Comment by endolith** [comment, marked as answer]
Apparently the proxy server isn't necessary.  @mariush444 [found a much simpler way][1]: Set the `.metainfo` file like this:

```
[url_template]
https://elevation.nationalmap.gov/arcgis/services/3DEPElevation/ImageServer/WMSServer?bbox={bbox}&format=image/jpeg&service=WMS&version=1.3.0&request=GetMap&CRS=CRS:84&width=256&height=256&layers=3DEPElevation:Hillshade%20Elevation%20Tinted&TRANSPARENT=TRUE&map=&styles=
[ext]
.jpg
[min_zoom]
1
[max_zoom]
19
[tile_size]
256
[img_density]
32
[avg_img_size]
18000
img_size]
18000
```

This is similar to the URLs I had been trying before, but it actually works.  From my testing it's due entirely to the use of `CRS:84` (which is in degrees and long, lat order?) instead of `EPSG:3857` (which is in meters and lat, long order?)

Anyway, I was able to get all the map layers into Osmand this way, without a proxy server, by replacing `Hillshade%20Elevation%20Tinted` with the other layer names:

- `Hillshade%20Gray`
  - "Provides a hillshaded surface generated dynamically using the hillshade server-side function on the elevation layer with a solar azimuth of 315 degrees and solar altitude of 45 degrees."
- `Aspect%20Map`
  - "This server-side function will apply an on-the-fly process to the input elevation data to generate a color visualization of aspect."
- `Hillshade%20Elevation%20Tinted`
  - "This server-side function will apply an on-the-fly process to the input elevation data to generate a tinted hillshade."
- `GreyHillshade_elevationFill`
  - "Provides a hillshaded surface generated dynamically using the hillshade server-side function on the elevation layer, with the voids filled in. A solar azimuth of 315 degrees and solar altitude of 45 degrees has been applied."
- `Hillshade%20Multidirectional`
  - "A raster function template."
- `Slope%20Map`
  - "This server-side function will apply an on-the-fly process to the input elevation data to generate a color visualization of slope, where flat surfaces are gray, shallow slopes are yellow, and steep slopes are red-brown."

[1]: https://github.com/osmandapp/OsmAnd/issues/4201#issuecomment-1702972896
